pippy/hf/_SaveModule.py

```diff
@@ -81,12 +86,29 @@ def _save_index(
             old_name = submod.remap_qualname(param_name)  # type: ignore
 
             binary_filename = _get_binary_filename(idx)
+
+            file_to_param_names[binary_filename].append(old_name)
+            # if `param_name` was previously mapped to another binary_filename
+            # and now it's changing, remove the previous mapping from `file_to_param_names`
+            if old_name in weight_map:
+                fn = weight_map.get(old_name)
+                if len(file_to_param_names[fn]) == 1:  # type: ignore
+                    del file_to_param_names[fn]  # type: ignore
+                else:
+                    file_to_param_names[fn].remove(  # type: ignore
+                        old_name
+                    )  # slow, find better way
+
+            #  add ckpt size once
+            if old_name not in weight_map:
+                index_dict["metadata"]["total_size"] += _get_param_size(param)  # type: ignore
+
             weight_map[old_name] = binary_filename
 
-            total_size += _get_param_size(param)
+            state_dict[old_name] = param
 
+    _save_weights(state_dict, file_to_param_names, checkpoint_dir)
```

Comment:
I think we want to do _save_weights in a separate function. The main reason behind it is that for the index file, we are currently just generating the index file on rank 0. However, for the weights, we definitely need to do it on every single rank.